src/main.rs

```diff
@@ -0,0 +1,552 @@
+
+    let epochs = config.first_epoch..=config.last_epoch;
+    loop {
+        // Pre-calculate the next_epoch_time for the InfoResponse hander.
+        let now = time::OffsetDateTime::now_utc();
+        let next_rotation = now + interval;
+        // Truncate to the nearest second.
+        let next_rotation = next_rotation
+            .replace_millisecond(0)
+            .expect("should be able to round to a fixed ms.");
+        let timestamp = next_rotation
+            .format(&Rfc3339)
+            .expect("well-known timestamp format should always succeed");
+        {
+            // Acquire a temporary write lock which should be dropped
+            // before sleeping. The locking should not fail, but if it
+            // does we can't set the field back to None, so panic rather
+            // than report stale information.
+            let mut s = state
+                .write()
+                .expect("should be able to update next_epoch_time");
+            s.next_epoch_time = Some(timestamp);
+        }
+
+        // Wait until the current epoch ends.
+        tokio::time::sleep(interval).await;
+
+        // Acquire exclusive access to the oprf state.
+        // Panics if this fails, since processing requests with an
+        // expired epoch weakens user privacy.
+        let mut s = state.write().expect("Failed to lock OPRFState");
+
+        // Puncture the current epoch so it can no longer be used.
+        let old_epoch = s.epoch;
+        s.server
+            .puncture(old_epoch)
+            .expect("Failed to puncture current epoch");
+
+        // Advance to the next epoch.
+        let new_epoch = old_epoch + 1;
+        if epochs.contains(&new_epoch) {
+            // Server is already initialized for this one.
+            s.epoch = new_epoch;
+        } else {
+            info!("Epochs exhausted! Rotating OPRF key");
+            // Panics if this fails. Puncture should mean we can't
+            // violate privacy through further evaluations, but we
+            // still want to drop the inner state with its private key.
+            *s = OPRFServer::new(config)
+                .expect("Could not initialize new PPOPRF state");
+        }
+        info!("epoch now {}", s.epoch);
+    }
+}
+
+/// Initialize an axum::Router for our web service
+/// Having this as a separate function makes testing easier.
+fn app(oprf_state: OPRFState) -> Router {
+    Router::new()
+        // Friendly default route to identify the site
+        .route("/", get(|| async { "STAR randomness server\n" }))
+        // Main endpoints
+        .route("/randomness", post(randomness))
+        .route("/info", get(info))
+        // Attach shared state
+        .with_state(oprf_state)
+        // Logging must come after active routes
+        .layer(tower_http::trace::TraceLayer::new_for_http())
+}
+
+/// Command line switches
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Config {
```

Comment:
adding brief help strings to each arg could be helpful for users

Reply:
Beyond what's already passed through the doc comments?

Reply:
yes, that way users of the CLI can quickly understand what each arg is (via the `--help` switch) for without peeking at the code